compiler/src/iree/compiler/Codegen/Common/CombineLayoutTransformation.cpp

```diff
@@ -442,6 +442,52 @@ insertIdentityMapScatter(RewriterBase &rewriter,
 LogicalResult
 combineLayoutTransformation(MLIRContext *ctx, FunctionOpInterface funcOp,
                             PadDistributionConfigFn padDistributionConfigFn) {
+  // Sink relayout operations to the end of the funcOp.
+  RewritePatternSet propagationPatterns(ctx);
+  tensor::populateFoldTensorEmptyPatterns(propagationPatterns);
+  tensor::ExpandShapeOp::getCanonicalizationPatterns(propagationPatterns, ctx);
+  tensor::CollapseShapeOp::getCanonicalizationPatterns(propagationPatterns,
+                                                       ctx);
+  // Only sink reshape ops, so bail if the consumer operation is a reshape.
+  auto controlSinkReshapesFn = [](OpOperand *operand) -> bool {
+    Operation *consumer = operand->getOwner();
+    return !llvm::isa<tensor::ExpandShapeOp, tensor::CollapseShapeOp>(consumer);
+  };
+  linalg::populateFoldReshapeOpsByExpansionPatterns(propagationPatterns,
+                                                    controlSinkReshapesFn);
+  // Only sink unpack ops, so bail if the producer operation is not an unpack.
+  // Also only sink unpack ops when new pack operations will not be created.
+  // This means the consumer op must have at most one additional destination
+  // operand, and it must come from an empty tensor.
+  auto controlPropagationFn = [](OpOperand *operand) -> bool {
+    if (!operand->get().getDefiningOp<linalg::UnPackOp>()) {
+      return false;
+    }
```

Comment:
IMO, it is easier to follow and contribute to the control function in the future, if you explicitly declare `producer/consumer` in the start of the function. E.g., I'll need to find other codes to understand which one is consumer and which one is producer.


```suggestion
  auto controlPropagationFn = [](OpOperand *operand) -> bool     Operation *producer = opOperand->get().getDefiningOp();
    Operation *consumer = opOperand->getOwner();
    if (!isa<linalg::UnPackOp>(producer)) {
      return false;
    }
```